// @devvit/client/effects/web-view-mode.js
import { WebViewImmersiveMode } from '@devvit/protos/json/devvit/ui/effects/web_view/v1alpha/immersive_mode.js';
import { emitEffect } from '@devvit/shared-types/client/emit-effect.js';
const modeListeners = new Set();
/**
 * Represents the current web view mode state for the application.
 *
 * @defaultValue 'inline'
 */
export const getWebViewMode = () => webViewMode(devvit.webViewMode);
/**
 * Requests expanded mode for the web view.
 * This will display the web view in a larger modal presentation.
 *
 * @param event The event that triggered the request, must be a trusted event.
 * @returns A promise that resolves request has been received.
 *
 * @example
 * ```ts
 * button.addEventListener('click', async (event) => {
 *   await requestExpandedMode(event);
 * });
 * ```
 */
export function requestExpandedMode(event) {
    return emitModeEffect(WebViewImmersiveMode.IMMERSIVE_MODE, event);
}
/**
 * Exits expanded mode for the web view.
 * This will display the web view in an inline presentation.
 *
 * @param event The event that triggered the request, must be a trusted event.
 * @returns A promise that resolves request has been received.
 *
 * @example
 * ```ts
 * button.addEventListener('click', async (event) => {
 *   await exitExpandedMode(event);
 * });
 * ```
 */
export function exitExpandedMode(event) {
    return emitModeEffect(WebViewImmersiveMode.INLINE_MODE, event);
}
/**
 * Adds a listener that is called when the web view mode changes.
 * The listener will be called with the new mode, either 'inline' or 'expanded'.
 *
 * @param callback The callback to be called when the mode changes.
 */
export function addWebViewModeListener(callback) {
    modeListeners.add(callback);
}
/**
 * Removes a listener that was previously added with `addWebViewModeListener`.
 *
 * @param callback The callback to be removed.
 */
export function removeWebViewModeListener(callback) {
    modeListeners.delete(callback);
}
async function emitModeEffect(mode, event) {
    if (!event.isTrusted || event.type !== 'click') {
        console.error('Expanded mode effect ignored due to untrusted event');
        throw new Error('Untrusted event');
    }
    const type = 9;
    await emitEffect({ type, immersiveMode: { immersiveMode: mode } });
}
/**
 * @internal
 * Handles incoming messages from the client, like when the user closes the immersive modal
 */
export function registerListener() {
    console.log('Registering web view mode listener');
    addEventListener('message', (event) => {
        const { type, data } = event.data;
        if (type !== 'devvit-message') {
            return;
        }
        if (!data?.immersiveModeEvent) {
            return;
        }
        devvit.webViewMode = data.immersiveModeEvent.immersiveMode;
        const webViewModeString = webViewMode(data.immersiveModeEvent.immersiveMode);
        modeListeners.forEach((listener) => listener(webViewModeString));
    });
}
function webViewMode(mode) {
    switch (mode) {
        case WebViewImmersiveMode.IMMERSIVE_MODE:
            return 'expanded';
        case WebViewImmersiveMode.INLINE_MODE:
        case WebViewImmersiveMode.UNRECOGNIZED:
        case WebViewImmersiveMode.UNSPECIFIED:
        case undefined:
            return 'inline';
        default:
            mode;
            throw Error(`${mode} not a WebViewImmersiveMode`);
    }
}
